products/llbuildAnalysis/CriticalPath.swift

```diff
@@ -0,0 +1,101 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+
+// The Swift package has llbuildSwift as module
+#if SWIFT_PACKAGE
+import llbuild
+import llbuildSwift
+#else
+import llbuild
+#endif
+
+/// Critical Path is a single linked list containing the first task built and pointers to dependents
+/// where the current element was the most expensive (long running) dependency of `next`.
+/// It conforms to Sequence providing access to all `BuildKey`s and their results in order.
+public final class CriticalPath: CustomStringConvertible {
+    public enum Error: Swift.Error, Equatable, CustomStringConvertible {
+        case missingDependency(self: CriticalPath, end: CriticalPath)
+        
+        public var description: String {
+            switch self {
+            case let .missingDependency(self, end):
+                return "Can't append \(end) to \(self) since \(self.key) is not a dependency of \(end.key). Found dependencies: \(end.result.dependencies)."
+            }
+        }
+    }
+    
+    /// The key that was built
+    public let key: BuildKey
+    /// The result containing the value, timing and dependency information
+    public let result: RuleResult
+    /// The cost of the whole path (sum up durations of all keys downstream)
+    public let cost: Double
+    /// The realistic cost of a critical path includes the waiting times between
+    /// the individual tasks.
+    public let realisticCost: Double
```

Comment:
I don’t know what this means, and haven’t grokked the code yet. I wonder if this isn’t something better handled elsewhere though

Reply:
(For example in a simulation tool). Critical path has a clear mathematical definition; I don’t know if this does as well

Reply:
Yeah, this is not really part of the mathematical calculation but a measurement that helps us to reason about the build. If the difference between this and the `cost` gets higher it means a more powerful machine would have built faster since tasks needed to wait although they were ready to go.

Reply:
On the other hand - it's easily calculatable afterwards with the start and end properties. I just thought it's easier to do that while building the graph and saving on the O(n) _(per graph)_ operations afterwards.

Reply:
I doubt that cost matters much in the scheme of things, I think clarity of the code is worth a lot more (see also comment about generic critical path algorithm)